### ISSUE TITLE
router.asPath shows previous URL on client side from _app.js getInitialProps

### ISSUE THREAD

**Issue by rashidul0405** [zeit/next.js#4369]
`console.log(router)` shows correct asPath but If I access `console.log(router.asPath)` shows the previous URL

- [x] I have searched the [issues](https://github.com/zeit/next.js/issues?q=is%3Aissue) of this repository and believe that this is not a duplicate.

## Code (`_app.js`)
```js
export default class MyApp extends App {
  constructor(props) {
    super(props)
  }

  static async getInitialProps({ Component, router, ctx }) {
    // Router.asPath shows current URL, can be seen from browser console as below
    console.log(router)
    // But if i access the asPath property as below, it shows previous URL. On server side, it's fine
    console.log(router.asPath)
  }
  ...
}
```

## Expected Behavior

Refer to above code, `router.asPath` should return current URL on client side

## Current Behavior

Refer to above code, `router.asPath` returns previous URL on client side

## Steps to Reproduce (for bugs)

Please refer to above code example.

## Context

After url is [deprecated](https://github.com/zeit/next.js/blob/master/errors/url-deprecated.md), we pass URL from `_app.js`, as all pages need this for some sort of SEO related stuffs.

as a workaround using `ctx.asPath` for now

## Your Environment

| Tech    | Version |
|---------|---------|
| next    |    6.0.0     |
| node    |    9.x     |
| OS      |    mac     |
| browser |    Chrome      |

**Comment by timneutkens** [comment]
The router is stateful by design at the moment. I do have some plans to change this in the future, but the recommended way right now is to use `ctx.asPath` like you said.

**Comment by rashidul0405** [comment]
Great, Thanks : )

**Comment by ex3ndr** [comment]
Sorry, what do you mean under stateful? How it affects? Should we everywhere use `ctx.asPath` instead of `router.asPath`? Can route can be updated before actual page transition (eg. old page get new URL first)?

**Comment by timneutkens** [comment]
Should be fixed in `next@canary`

**Comment by timneutkens** [comment]
Specifically this PR fixed it: #4541
